tests/test_interface.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This file is part of ansible-nmstate.
+#
+# ansible-nmstate is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ansible-nmstate is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ansible-nmstate.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+""" unit tests for nmstate_linkagg.py """
+
+try:
+    from unittest import mock
+except ImportError:  # py2
+    import mock
+
+import sys
+
+sys.modules["ansible"] = mock.Mock()
+sys.modules["ansible.module_utils.basic"] = mock.Mock()
+sys.modules["ansible.module_utils"] = mock.Mock()
+sys.modules["ansible.module_utils.network.common"] = mock.Mock()
+sys.modules["ansible.module_utils.network.common.utils"] = mock.Mock()
+sys.modules["ansible.module_utils.network"] = mock.Mock()
+
+sys.modules["ansible.module_utils.ansible_nmstate"] = __import__("ansible_nmstate")
+
+import nmstate_interface as nif  # noqa: E402
+
+
+def test_import_succeeded():
+    assert nif
+
+
+def test_all_settings(base_state):
+    ansible_module = mock.Mock()
+    ansible_module.check_mode = False
+    ansible_module.params = dict(
+        name="eth0",
+        state="present",
+        enabled="yes",
+        delay="10",
+        mtu="1500",
+        duplex="full",
+        description="test_description",
+        speed="1000",
+    )
+
+    def side_effect():
+        return base_state
+
+    with mock.patch(
+        "ansible_nmstate.netinfo.show", side_effect=side_effect
+    ) as mock_show:
+        nmstate_module = nif.AnsibleNMStateInterface(
+            ansible_module, "nmstate_interface_mock"
+        )
+    assert mock_show.called
+
+    with mock.patch(
+        "ansible_nmstate.netinfo.show", side_effect=side_effect
+    ) as mock_show:
+        with mock.patch("ansible_nmstate.netapplier.apply") as mock_apply:
+            nmstate_module.run()
+
+    assert mock_show.called
+    assert mock_apply.called
+    applied_state = mock_apply.call_args[0][0]
+
+    interface_state = applied_state["interfaces"][0]
+    assert interface_state["name"] == "eth0"
```

Comment:
I would to assert only the single settings that the module handles. A full dict equality would also have to assert the device type (and possible other settings) that are not relevant.

Reply:
You can normalize the dict for the assertion.
It can help and see all the data when one fails. With multiple assertions we do not see the others.

Reply:
What do you mean with normalization? Removing all the items that I do not care about?

Reply:
Yes, building a dict for assertion: On one hand normalize a copy of interface_state and on the other build the expected dict.

Reply:
This seems to be very error-prone since it will require a lot of extra, untested code that hopefully does the right thing. Also it makes it harder to assert that values are not missing. This is a possible implementation but it does not convince me:

```
diff --git a/tests/test_interface.py b/tests/test_interface.py
index 5c7844d..92ed615 100644
--- a/tests/test_interface.py
+++ b/tests/test_interface.py
@@ -23,7 +23,7 @@ try:
     from unittest import mock
 except ImportError:  # py2
     import mock
-
+from copy import deepcopy
 
 mock_modules = {}
 mock_modules["ansible"] = mock.Mock()
@@ -41,6 +41,24 @@ with mock.patch.dict("sys.modules", mock_modules):
     import nmstate_interface as nif  # noqa: E402
 
 
+def whitelist_keys(whitelist, d):
+    result = {}
+    if isinstance(whitelist, dict):
+        for k, v in d.items():
+            if k in whitelist:
+                if isinstance(v, dict):
+                    if k in whitelist:
+                        result[k] = whitelist_keys(whitelist[k], v)
+                elif isinstance(v, list):
+                    clean = []
+                    for i in range(len(v)):
+                        clean.append(whitelist_keys(whitelist[k][i], v[i]))
+                    result[k] = clean
+                else:
+                    result[k] = v
+    return result
+
+
 def test_import_succeeded():
     assert nif
 
@@ -86,3 +104,31 @@ def test_all_settings(base_state):
     assert ethernet_state["auto-negotiation"] is False
     assert ethernet_state["duplex"] == "full"
     assert ethernet_state["speed"] == 1000
+
+    expected_state = {
+        "interfaces": [
+            {
+                "name": "eth0",
+                "description": "test_description",
+                "mtu": 1500,
+                "ethernet": {
+                    "auto-negotiation": False,
+                    "duplex": "full",
+                    "speed": 1000,
+                },
+            }
+        ]
+    }
+
+    whitelist = deepcopy(expected_state)
+
+    # add state to whitelist to assert that it is included
+    whitelist["state"] = None
+    normalized_output = whitelist_keys(whitelist, applied_state)
+    assert normalized_output == expected_state
```

Is there maybe a way to make pytest just show the full dictionary if the assertion of a key fails? However, I am also not sure if I really need the full state on failures...

Reply:
You can make pytest should you something specific: `assert a == b, '{} != {}'.format(x, y)`

But you can also take this approach which is simpler:
assert {"name": interface_state["name"], ...} == {"name": "eth0", ...}